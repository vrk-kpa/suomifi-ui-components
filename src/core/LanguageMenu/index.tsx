export { LanguageMenu, type LanguageMenuProps } from './LanguageMenu';
export {
  LanguageMenuItem,
  type LanguageMenuItemProps,
} from './LanguageMenuItem/LanguageMenuItem';
