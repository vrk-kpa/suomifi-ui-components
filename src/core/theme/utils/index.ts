export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
