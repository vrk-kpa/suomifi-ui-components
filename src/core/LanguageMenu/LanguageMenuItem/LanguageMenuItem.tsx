import React, { ReactNode } from 'react';
import classnames from 'classnames';
import { HtmlButton } from '../../../reset';
import { SuomifiThemeConsumer, SuomifiThemeProp } from '../../theme';
import { baseStyles } from './LanguageMenuItem.baseStyles';
import { styled } from 'styled-components';
import {
  LanguageMenuProviderState,
  LanguageMenuConsumer,
} from './../LanguageMenuPopover/LanguageMenuPopover';

export interface LanguageMenuItemProps {
  /** Language code for language menu item. Must be provided for assistive technology. */
  lang: string;
  /** CSS class for custom styles */
  className?: string;
  /** Text for the language menu item. Text should contain full language name and not only the abbreviation. */
  children: ReactNode;
  /** Callback fired when menu item is clicked */
  onSelect: () => void;
  /** Marks the item as the currently selected one */
  selected?: boolean;
}

interface BaseLanguageMenuItemProps extends LanguageMenuItemProps {
  consumer: LanguageMenuProviderState;
  itemIndex?: number; // Index number of the child. For internal use only. Added by LanguageMenuPopover.
}

const baseClassName = 'fi-language-menu-item';
const hasKeyboardFocusClassName = `${baseClassName}--isHighlighted`;
const selectedClassName = `${baseClassName}--selected`;

const BaseLanguageMenuItem = (
  props: BaseLanguageMenuItemProps & SuomifiThemeProp,
) => {
  const {
    className,
    children,
    selected,
    onSelect,
    consumer,
    itemIndex = -1,
    theme,
    ...passProps
  } = props;

  const hasKeyboardFocus = consumer.activeDescendantIndex === itemIndex;
  const listElementRef = React.useRef<HTMLButtonElement>(null);

  React.useLayoutEffect(() => {
    if (hasKeyboardFocus) {
      listElementRef.current?.focus();
    }
  }, [consumer.activeDescendantIndex]);

  return (
    <HtmlButton
      forwardedRef={listElementRef}
      className={classnames(baseClassName, className, {
        [hasKeyboardFocusClassName]: hasKeyboardFocus,
        [selectedClassName]: selected,
      })}
      onClick={() => {
        onSelect();
        consumer.onItemClick(itemIndex);
      }}
      onMouseOver={() => {
        consumer.onItemMouseOver(itemIndex);
      }}
      onKeyDown={(event) => {
        if (event.key === 'Enter' || event.key === ' ') {
          event.preventDefault();
          listElementRef.current?.click();
        }
      }}
      tabIndex={hasKeyboardFocus ? 0 : -1}
      id={`${consumer.parentId}-list-item-${itemIndex}`}
      role="menuitem"
      aria-current={selected}
      {...passProps}
    >
      {children}
    </HtmlButton>
  );
};

const StyledLanguageMenuItem = styled(BaseLanguageMenuItem)`
  ${({ theme }) => baseStyles(theme)}
`;

const LanguageMenuItem = (props: LanguageMenuItemProps) => (
  <SuomifiThemeConsumer>
    {({ suomifiTheme }) => (
      <LanguageMenuConsumer>
        {(consumer) => (
          <StyledLanguageMenuItem
            theme={suomifiTheme}
            consumer={consumer}
            {...props}
          />
        )}
      </LanguageMenuConsumer>
    )}
  </SuomifiThemeConsumer>
);

LanguageMenuItem.displayName = 'LanguageMenuItem';
export { LanguageMenuItem };
