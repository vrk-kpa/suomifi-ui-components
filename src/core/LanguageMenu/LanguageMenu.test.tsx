import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import { axeTest } from '../../utils/test';
import { LanguageMenu, LanguageMenuProps } from './LanguageMenu';
import { LanguageMenuItem } from './LanguageMenuItem/LanguageMenuItem';
import { HTMLAttributesIncludingDataAttributes } from 'utils/common/common';

type MenuProps = LanguageMenuProps &
  HTMLAttributesIncludingDataAttributes<HTMLButtonElement>;

const languageMenuProps: MenuProps = {
  'aria-label': 'Select language',
  buttonText: 'In English (EN)',
};

const mockOnSelect = jest.fn();

const TestLanguageMenu = (props: MenuProps) => (
  <LanguageMenu {...props}>
    <LanguageMenuItem onSelect={mockOnSelect} lang="fi">
      Suomeksi (FI)
    </LanguageMenuItem>
    <LanguageMenuItem onSelect={() => ({})} lang="sv">
      PÃ¥ svenska (SV)
    </LanguageMenuItem>
    <LanguageMenuItem onSelect={() => ({})} lang="en" selected>
      In English (EN)
    </LanguageMenuItem>
  </LanguageMenu>
);

describe('Basic LanguageMenu', () => {
  it('should match snapshot when closed', async () => {
    const { baseElement } = render(TestLanguageMenu(languageMenuProps));
    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot when opened', async () => {
    const { baseElement, getByRole } = render(
      TestLanguageMenu(languageMenuProps),
    );
    const menuButton = getByRole('button');
    await act(async () => {
      fireEvent.click(menuButton);
      await new Promise((resolve) => {
        setTimeout(resolve, 100);
      });
    });
    expect(baseElement).toMatchSnapshot();
  });
});

describe('LanguageMenuItem', () => {
  describe('onSelect', () => {
    it('should call onSelect when clicked', () => {
      const { getByRole, getAllByRole } = render(
        TestLanguageMenu(languageMenuProps),
      );
      fireEvent.click(getByRole('button'));
      const item = getAllByRole('menuitem')[0];
      fireEvent.click(item);
      expect(mockOnSelect).toHaveBeenCalledTimes(1);
    });
  });
});

describe('props', () => {
  describe('buttonText', () => {
    it('should have buttonText', async () => {
      const { findByRole } = render(TestLanguageMenu(languageMenuProps));
      const button = await findByRole('button');
      expect(button).toHaveTextContent('In English (EN)');
    });
  });

  describe('id', () => {
    it('should have id', () => {
      const { getByRole } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          id: 'test-id',
        }),
      );
      const button = getByRole('button');
      expect(button).toHaveAttribute('id', 'test-id');
    });
  });

  describe('data-testid', () => {
    it('should have data-testid', () => {
      const { getByRole } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          'data-testid': 'custom-data-attribute',
        }),
      );
      const button = getByRole('button');
      expect(button).toHaveAttribute('data-testid', 'custom-data-attribute');
    });
  });

  describe('className', () => {
    it('shoud have base className and given className in wrapper element', () => {
      const { container } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          className: 'lm-test',
        }),
      );
      expect(container.firstChild).toHaveClass('fi-language-menu');
      expect(container.firstChild).toHaveClass('lm-test');
    });
  });

  describe('menuClassName', () => {
    it('shoud have className in popover wrapper element', () => {
      const { baseElement } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          menuClassName: 'menu-custom-class',
        }),
      );
      const div = baseElement.querySelector('.fi-language-menu-popover');
      expect(div).toHaveClass('menu-custom-class');
    });
  });

  describe('margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          margin: 'xs',
        }),
      );
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop overwritten by style prop', () => {
      const { container } = render(
        TestLanguageMenu({
          ...languageMenuProps,
          margin: 'xs',
          style: { margin: 2 },
        }),
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });
});

describe('callbacks', () => {
  describe('onOpen', () => {
    it('should call onOpen when menu button is clicked', () => {
      const mockOnOpen = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onOpen: mockOnOpen }),
      );
      fireEvent.click(getByRole('button'));
      expect(mockOnOpen).toBeCalledTimes(1);
    });

    it('should call onOpen with Enter in menu button', () => {
      const mockOnOpen = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onOpen: mockOnOpen }),
      );
      fireEvent.keyDown(getByRole('button'), { key: 'Enter' });
      expect(mockOnOpen).toBeCalledTimes(1);
    });

    it('should call onOpen with Space in menu button', () => {
      const mockOnOpen = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onOpen: mockOnOpen }),
      );
      fireEvent.keyDown(getByRole('button'), { key: ' ' });
      expect(mockOnOpen).toBeCalledTimes(1);
    });

    it('should call onOpen with ArrowUp in menu button', () => {
      const mockOnOpen = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onOpen: mockOnOpen }),
      );
      fireEvent.keyDown(getByRole('button'), { key: 'ArrowUp' });
      expect(mockOnOpen).toBeCalledTimes(1);
    });

    it('should call onOpen with ArrowDown in menu button', () => {
      const mockOnOpen = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onOpen: mockOnOpen }),
      );
      fireEvent.keyDown(getByRole('button'), { key: 'ArrowDown' });
      expect(mockOnOpen).toBeCalledTimes(1);
    });
  });

  describe('onClose', () => {
    it('should call onClose with Escape', () => {
      const mockOnClose = jest.fn();
      const { getByRole, getAllByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onClose: mockOnClose }),
      );
      fireEvent.click(getByRole('button'));
      const item = getAllByRole('menuitem')[0];
      fireEvent.click(item);
      fireEvent.keyDown(item, { key: 'Escape' });
      expect(mockOnClose).toBeCalledTimes(1);
    });

    it('should call onClose with click outside menu', async () => {
      const mockOnClose = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onClose: mockOnClose }),
      );
      fireEvent.click(getByRole('button'));
      expect(getByRole('menu')).toBeVisible();
      fireEvent.click(getByRole('button'));
      expect(mockOnClose).toBeCalledTimes(1);
    });
  });

  describe('onBlur', () => {
    it('should call onBlur when focus moves to menu', () => {
      const mockOnBlur = jest.fn();
      const { getByRole } = render(
        TestLanguageMenu({ ...languageMenuProps, onBlur: mockOnBlur }),
      );
      fireEvent.blur(getByRole('button'));
      expect(mockOnBlur).toBeCalledTimes(1);
    });
  });
});

// Don't validate aria-attributes since Portal is not rendered and there is no pair for aria-controls
test(
  'should not have basic accessibility issues',
  axeTest(TestLanguageMenu(languageMenuProps), {
    rules: {
      'aria-valid-attr-value': {
        enabled: false,
      },
    },
  }),
);
