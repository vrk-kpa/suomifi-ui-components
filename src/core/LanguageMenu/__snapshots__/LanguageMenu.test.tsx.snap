// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic LanguageMenu should match snapshot when closed 1`] = `
.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c2:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c2::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c2::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c2::-webkit-inner-spin-button {
  height: auto;
}

.c2::-webkit-outer-spin-button {
  height: auto;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c5.fi-language-menu-popover {
  background-color: hsl(0,0%,100%);
  box-shadow: 0px 4px 8px 0px hsla(0,0%,13%,0.14);
  border: 1px solid hsl(0,0%,42%);
  border-radius: 2px;
  z-index: 888;
}

.c5.fi-language-menu-popover--hidden {
  visibility: hidden;
}

.c5 .fi-language-menu-popover_list {
  padding-top: 5px;
  padding-bottom: 5px;
  margin: 0;
  padding-left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  list-style-type: none;
  overflow-y: auto;
  max-height: 260px;
}

.c5 .fi-language-menu-popover_list:focus {
  outline: none;
}

.c5 .fi-language-menu-popover_popper-arrow::before,
.c5 .fi-language-menu-popover_popper-arrow::after {
  content: '';
  position: absolute;
  left: -9px;
  height: 0;
  width: 0;
  border: solid transparent;
  pointer-events: none;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end'] {
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end']::before {
  border-bottom-color: hsl(0,0%,42%);
  border-width: 9px;
  margin-right: -9px;
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end']::after {
  border-bottom-color: hsl(0,0%,100%);
  border-width: 8px;
  margin-right: -9px;
  left: -8px;
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end'] {
  top: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end']::before {
  border-top-color: hsl(0,0%,42%);
  border-width: 9px;
  margin-right: -9px;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end']::after {
  border-top-color: hsl(0,0%,100%);
  border-width: 8px;
  margin-right: -9px;
  left: -8px;
}

.c3 {
  vertical-align: baseline;
}

.c3.fi-icon {
  display: inline-block;
}

.c3 .fi-icon-base-fill {
  fill: currentColor;
}

.c3 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c3.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c3.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c1 {
  color: hsl(0,0%,13%);
}

.c1.fi-language-menu .fi-language-menu_button {
  color: hsl(0,0%,13%);
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  padding: 9px 10px;
  line-height: 24px;
  background-color: hsl(0,0%,100%);
  border: 1px solid transparent;
  border-radius: 2px;
  word-break: break-word;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_text_wrapper {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  word-break: break-word;
  line-height: 24px;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_icon {
  width: 18px;
  height: 18px;
  margin-left: 10px;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_icon .fi-icon-base-fill {
  fill: hsl(212,63%,45%);
}

.c1.fi-language-menu .fi-language-menu_button:focus {
  position: relative;
  outline: 3px solid transparent;
}

.c1.fi-language-menu .fi-language-menu_button:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1.fi-language-menu .fi-language-menu_button:hover {
  border-color: hsl(202,7%,80%);
}

.c1.fi-language-menu .fi-language-menu_button--open {
  border-color: hsl(202,7%,80%);
}

.c6.fi-language-menu-item {
  color: hsl(0,0%,13%);
  margin: 5px 0;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  word-break: break-word;
  color: hsl(0,0%,13%);
  display: inline-block;
  width: 100%;
  padding: 0 20px 0 5px;
  border-left: 4px solid transparent;
}

.c6.fi-language-menu-item:active,
.c6.fi-language-menu-item:focus,
.c6.fi-language-menu-item:visited {
  outline: none;
  color: hsl(0,0%,13%);
}

.c6.fi-language-menu-item--isHighlighted {
  border-left-color: hsl(212,63%,45%);
  position: relative;
}

.c6.fi-language-menu-item--isHighlighted:after {
  content: '';
}

.c6.fi-language-menu-item--selected {
  font-weight: bold;
  border-left-color: hsl(212,63%,45%);
}

@media (forced-colors:active) {
  .c6.fi-language-menu-item--isHighlighted:after {
    position: absolute;
    left: 1px;
    right: 1px;
    top: 1px;
    bottom: 1px;
    border: solid 3px Highlight;
  }
}

<body>
  <div>
    <div
      class="c0 fi-language-menu c1"
    >
      <button
        aria-controls="1-menu"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Select language"
        class="c2 fi-language-menu_button"
        id="1"
        type="button"
      >
        In English (EN)
        <svg
          aria-hidden="true"
          class="fi-icon c3 fi-language-menu_button_icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M12.99 15.61c-.273.26-.632.39-.99.39s-.717-.13-.99-.39l-5.6-5.334a1.287 1.287 0 0 1 0-1.885 1.448 1.448 0 0 1 1.98 0l4.61 4.39 4.61-4.39a1.448 1.448 0 0 1 1.98 0c.547.52.547 1.365 0 1.885l-5.6 5.333Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <div
        class="c4 c5 fi-language-menu-popover fi-language-menu-popover--hidden"
        style="position: fixed; left: 0px; top: 0px;"
      >
        <div
          class="c4 fi-language-menu-popover_list"
          id="1-menu"
          role="menu"
          tabindex="-1"
        >
          <button
            class="c2 fi-language-menu-item c6"
            id="1-list-item-0"
            lang="fi"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            Suomeksi (FI)
          </button>
          <button
            class="c2 fi-language-menu-item c6"
            id="1-list-item-1"
            lang="sv"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            PÃ¥ svenska (SV)
          </button>
          <button
            aria-current="true"
            class="c2 fi-language-menu-item c6 fi-language-menu-item--selected"
            id="1-list-item-2"
            lang="en"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            In English (EN)
          </button>
        </div>
        <div
          class="c0 fi-language-menu-popover_popper-arrow"
          data-popper-arrow="true"
          style="position: absolute;"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`Basic LanguageMenu should match snapshot when opened 1`] = `
.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c2:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c2::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c2::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c2::-webkit-inner-spin-button {
  height: auto;
}

.c2::-webkit-outer-spin-button {
  height: auto;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c5.fi-language-menu-popover {
  background-color: hsl(0,0%,100%);
  box-shadow: 0px 4px 8px 0px hsla(0,0%,13%,0.14);
  border: 1px solid hsl(0,0%,42%);
  border-radius: 2px;
  z-index: 888;
}

.c5.fi-language-menu-popover--hidden {
  visibility: hidden;
}

.c5 .fi-language-menu-popover_list {
  padding-top: 5px;
  padding-bottom: 5px;
  margin: 0;
  padding-left: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  list-style-type: none;
  overflow-y: auto;
  max-height: 260px;
}

.c5 .fi-language-menu-popover_list:focus {
  outline: none;
}

.c5 .fi-language-menu-popover_popper-arrow::before,
.c5 .fi-language-menu-popover_popper-arrow::after {
  content: '';
  position: absolute;
  left: -9px;
  height: 0;
  width: 0;
  border: solid transparent;
  pointer-events: none;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end'] {
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end']::before {
  border-bottom-color: hsl(0,0%,42%);
  border-width: 9px;
  margin-right: -9px;
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='bottom-end']::after {
  border-bottom-color: hsl(0,0%,100%);
  border-width: 8px;
  margin-right: -9px;
  left: -8px;
  bottom: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end'] {
  top: 100%;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end']::before {
  border-top-color: hsl(0,0%,42%);
  border-width: 9px;
  margin-right: -9px;
}

.c5 .fi-language-menu-popover_popper-arrow[data-popper-placement^='top-end']::after {
  border-top-color: hsl(0,0%,100%);
  border-width: 8px;
  margin-right: -9px;
  left: -8px;
}

.c3 {
  vertical-align: baseline;
}

.c3.fi-icon {
  display: inline-block;
}

.c3 .fi-icon-base-fill {
  fill: currentColor;
}

.c3 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c3.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c3.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c1 {
  color: hsl(0,0%,13%);
}

.c1.fi-language-menu .fi-language-menu_button {
  color: hsl(0,0%,13%);
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  padding: 9px 10px;
  line-height: 24px;
  background-color: hsl(0,0%,100%);
  border: 1px solid transparent;
  border-radius: 2px;
  word-break: break-word;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_text_wrapper {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  word-break: break-word;
  line-height: 24px;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_icon {
  width: 18px;
  height: 18px;
  margin-left: 10px;
}

.c1.fi-language-menu .fi-language-menu_button .fi-language-menu_button_icon .fi-icon-base-fill {
  fill: hsl(212,63%,45%);
}

.c1.fi-language-menu .fi-language-menu_button:focus {
  position: relative;
  outline: 3px solid transparent;
}

.c1.fi-language-menu .fi-language-menu_button:focus:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c1.fi-language-menu .fi-language-menu_button:hover {
  border-color: hsl(202,7%,80%);
}

.c1.fi-language-menu .fi-language-menu_button--open {
  border-color: hsl(202,7%,80%);
}

.c6.fi-language-menu-item {
  color: hsl(0,0%,13%);
  margin: 5px 0;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  word-break: break-word;
  color: hsl(0,0%,13%);
  display: inline-block;
  width: 100%;
  padding: 0 20px 0 5px;
  border-left: 4px solid transparent;
}

.c6.fi-language-menu-item:active,
.c6.fi-language-menu-item:focus,
.c6.fi-language-menu-item:visited {
  outline: none;
  color: hsl(0,0%,13%);
}

.c6.fi-language-menu-item--isHighlighted {
  border-left-color: hsl(212,63%,45%);
  position: relative;
}

.c6.fi-language-menu-item--isHighlighted:after {
  content: '';
}

.c6.fi-language-menu-item--selected {
  font-weight: bold;
  border-left-color: hsl(212,63%,45%);
}

@media (forced-colors:active) {
  .c6.fi-language-menu-item--isHighlighted:after {
    position: absolute;
    left: 1px;
    right: 1px;
    top: 1px;
    bottom: 1px;
    border: solid 3px Highlight;
  }
}

<body>
  <div>
    <div
      class="c0 fi-language-menu c1"
    >
      <button
        aria-controls="3-menu"
        aria-expanded="true"
        aria-haspopup="menu"
        aria-label="Select language"
        class="c2 fi-language-menu_button fi-language-menu_button--open"
        id="3"
        type="button"
      >
        In English (EN)
        <svg
          aria-hidden="true"
          class="fi-icon c3 fi-language-menu_button_icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M12.99 8.39A1.433 1.433 0 0 0 12 8c-.358 0-.717.13-.99.39l-5.6 5.334a1.287 1.287 0 0 0 0 1.885 1.448 1.448 0 0 0 1.98 0l4.61-4.39 4.61 4.39a1.448 1.448 0 0 0 1.98 0 1.287 1.287 0 0 0 0-1.885l-5.6-5.333Z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <div
        class="c4 c5 fi-language-menu-popover"
        style="position: fixed; left: 0px; top: 0px; right: 0px; transform: translate(0px, 15px);"
      >
        <div
          class="c4 fi-language-menu-popover_list"
          id="3-menu"
          role="menu"
          tabindex="-1"
        >
          <button
            class="c2 fi-language-menu-item c6 fi-language-menu-item--isHighlighted"
            id="3-list-item-0"
            lang="fi"
            role="menuitem"
            tabindex="0"
            type="button"
          >
            Suomeksi (FI)
          </button>
          <button
            class="c2 fi-language-menu-item c6"
            id="3-list-item-1"
            lang="sv"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            PÃ¥ svenska (SV)
          </button>
          <button
            aria-current="true"
            class="c2 fi-language-menu-item c6 fi-language-menu-item--selected"
            id="3-list-item-2"
            lang="en"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            In English (EN)
          </button>
        </div>
        <div
          class="c0 fi-language-menu-popover_popper-arrow"
          data-popper-arrow="true"
          data-popper-placement="bottom-end"
          style="position: absolute; left: 0px; transform: translate(0px, 0px);"
        />
      </div>
    </div>
  </div>
</body>
`;
